File: ml-controller/services/alpha_model_roster.py
"""Immutable eight-model identities; TimeXer replaces, never aliases, DLinear."""
from __future__ import annotations

LEGACY_MODELS = ("LightGBM", "XGBoost", "ExtraTrees", "TabM", "GNN",
                 "DLinear", "PatchTST", "iTransformer")
TIMEXER_MODELS = tuple("TimeXer" if name == "DLinear" else name for name in LEGACY_MODELS)
SUPPORTED_MODELS = frozenset((*LEGACY_MODELS, "TimeXer"))


def model_order(names, *, complete=False) -> tuple[str, ...]:
    names = set(names)
    if "DLinear" in names and "TimeXer" in names:
        raise ValueError("alpha_roster_mixed_replacement_slot")
    order = TIMEXER_MODELS if "TimeXer" in names else LEGACY_MODELS
    if complete and names != set(order):
        raise ValueError("alpha_roster_requires_exactly_eight_models")
    return order


def validate_order(names) -> tuple[str, ...]:
    order = tuple(names)
    if order not in (LEGACY_MODELS, TIMEXER_MODELS):
        raise ValueError("alpha_roster_order_invalid")
    return order


def stacker_features(order) -> tuple[str, ...]:
    order = validate_order(order)
    return tuple([f"{name}.rank" for name in order]
                 + [f"{name}.available" for name in order])


def is_complete_roster(names) -> bool:
    try:
        model_order(names, complete=True)
        return True
    except (ValueError, TypeError):
        return False


def published_model_order(query):
    """The ensemble pointer owns the slot replacement, never newest artifacts."""
    import json
    rows = query("SELECT a.payload_json, a.payload_checksum, a.state, a.production_effect, "
        "p.payload_checksum AS pointer_payload_checksum FROM active8_ensemble_pointer_v1 p "
        "JOIN active8_ensemble_artifacts_v1 a ON a.artifact_id=p.artifact_id WHERE p.singleton_id=1", [])
    if not rows:
        return LEGACY_MODELS
    if len(rows) != 1:
        raise ValueError('alpha_roster_published_pointer_ambiguous')
    row = rows[0]
    payload = json.loads(row['payload_json'])
    if (row.get('state') != 'production' or int(row.get('production_effect') or 0) != 1
            or row.get('payload_checksum') != row.get('pointer_payload_checksum')
            or payload.get('payload_checksum') != row.get('payload_checksum')):
        raise ValueError('alpha_roster_published_pointer_invalid')
    return validate_order(payload['model_order'])

File: worker/src/lib/alphaModelRoster.ts
/** Published eight-model identity; a research artifact never replaces a slot. */
export const LEGACY_ALPHA_MODELS = ['LightGBM', 'XGBoost', 'ExtraTrees', 'TabM', 'GNN', 'DLinear', 'PatchTST', 'iTransformer'] as const
export const TIMEXER_ALPHA_MODELS = LEGACY_ALPHA_MODELS.map(name => name === 'DLinear' ? 'TimeXer' : name)
export function validateAlphaModelOrder(value: unknown): string[] {
  if (!Array.isArray(value) || ![LEGACY_ALPHA_MODELS, TIMEXER_ALPHA_MODELS].some(order => order.length === value.length && order.every((name, i) => name === value[i]))) throw new Error('alpha_roster_requires_exactly_eight_models')
  return [...value]
}
export function predictionModelOrder(data: any): string[] {
  const declared = data?.ensemble_v2?.model_order ?? data?.active_model_contract
  return declared == null ? [...LEGACY_ALPHA_MODELS] : validateAlphaModelOrder(declared)
}
export async function publishedAlphaModelOrder(db: D1Database): Promise<string[]> {
  const row = await db.prepare(`SELECT a.payload_json, a.state, a.production_effect, a.payload_checksum, p.payload_checksum AS pointer_checksum FROM active8_ensemble_pointer_v1 p LEFT JOIN active8_ensemble_artifacts_v1 a ON a.artifact_id=p.artifact_id WHERE p.singleton_id=1`).first<{payload_json:string;state:string;production_effect:number;payload_checksum:string;pointer_checksum:string}>()
  if (!row) return [...LEGACY_ALPHA_MODELS]
  const payload = JSON.parse(row.payload_json)
  if (row.state !== 'production' || row.production_effect !== 1 || row.payload_checksum !== row.pointer_checksum || payload.payload_checksum !== row.payload_checksum) throw new Error('alpha_roster_published_pointer_invalid')
  return validateAlphaModelOrder(payload.model_order)
}
